File: frontend/src/types/observability.ts
export type ObservabilityRouteMetric = {
  route: string;
  requestCount: number;
  errorCount: number;
  slowCount: number;
  avgLatencyMs: number;
  p95LatencyMs: number;
  maxLatencyMs: number;
};

export type ObservabilityDomainMetric = {
  domain: string;
  requestCount: number;
  errorCount: number;
  slowCount: number;
};

export type ObservabilityErrorCodeMetric = {
  errorCode: string;
  count: number;
};

export type ObservabilityErrorEvent = {
  requestId: string;
  method: string;
  path: string;
  route: string;
  statusCode: number;
  durationMs: number;
  errorCode?: string | null;
  userId?: string | null;
  userRole?: string | null;
  occurredAt: string;
};

export type ObservabilityTaskFailureMetric = {
  jobId: string;
  failedCount: number;
  lastFailedAt?: string | null;
  lastError?: string | null;
};

export type ObservabilityOverviewPayload = {
  windowHours: number;
  slowThresholdMs: number;
  totalRequests: number;
  errorRequests: number;
  slowRequests: number;
  p95LatencyMs: number;
  topRoutes: ObservabilityRouteMetric[];
  domainStats: ObservabilityDomainMetric[];
  errorCodeStats: ObservabilityErrorCodeMetric[];
  taskFailures: ObservabilityTaskFailureMetric[];
};

export type ObservabilityErrorsPayload = {
  windowHours: number;
  items: ObservabilityErrorEvent[];
};

export type ObservabilitySlowRequestsPayload = {
  windowHours: number;
  slowThresholdMs: number;
  items: ObservabilityErrorEvent[];
};

export type ObservabilityTaskFailuresPayload = {
  windowHours: number;
  items: ObservabilityTaskFailureMetric[];
};

File: backend/app/schemas/observability.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class ObservabilityRouteMetric(BaseModel):
    route: str
    requestCount: int
    errorCount: int
    slowCount: int
    avgLatencyMs: float
    p95LatencyMs: float
    maxLatencyMs: float


class ObservabilityDomainMetric(BaseModel):
    domain: str
    requestCount: int
    errorCount: int
    slowCount: int


class ObservabilityErrorCodeMetric(BaseModel):
    errorCode: str
    count: int


class ObservabilityErrorEvent(BaseModel):
    requestId: str
    method: str
    path: str
    route: str
    statusCode: int
    durationMs: float
    errorCode: str | None = None
    userId: str | None = None
    userRole: str | None = None
    occurredAt: datetime


class ObservabilityTaskFailureMetric(BaseModel):
    jobId: str
    failedCount: int
    lastFailedAt: datetime | None = None
    lastError: str | None = None


class ObservabilityOverviewResponse(BaseModel):
    windowHours: int
    slowThresholdMs: int
    totalRequests: int
    errorRequests: int
    slowRequests: int
    p95LatencyMs: float
    topRoutes: list[ObservabilityRouteMetric] = Field(default_factory=list)
    domainStats: list[ObservabilityDomainMetric] = Field(default_factory=list)
    errorCodeStats: list[ObservabilityErrorCodeMetric] = Field(default_factory=list)
    taskFailures: list[ObservabilityTaskFailureMetric] = Field(default_factory=list)


class ObservabilityErrorsResponse(BaseModel):
    windowHours: int
    items: list[ObservabilityErrorEvent] = Field(default_factory=list)


class ObservabilitySlowRequestsResponse(BaseModel):
    windowHours: int
    slowThresholdMs: int
    items: list[ObservabilityErrorEvent] = Field(default_factory=list)


class ObservabilityTaskFailuresResponse(BaseModel):
    windowHours: int
    items: list[ObservabilityTaskFailureMetric] = Field(default_factory=list)
